@Likith101 it seems there're a few tests failing within `s2/`, although none that you've added it seems ..

```
ok  	github.com/blevesearch/geo/geojson	0.201s
ok  	github.com/blevesearch/geo/r1	0.827s
ok  	github.com/blevesearch/geo/r2	0.366s
ok  	github.com/blevesearch/geo/r3	0.520s
ok  	github.com/blevesearch/geo/s1	0.673s
--- FAIL: TestClosestEdgeQueryTrueDistanceLessThanChordAngleDistance (0.00s)
    edge_query_closest_test.go:122: CompareDistance((0.785167625848291916845767, -0.502004006908459698976799, -0.362634494177826782745910), (0.785630117324294330316548, -0.501876559404935029817807, -0.361808288839380542967206), 9.127564928066267e-07) = 1, want >= 0
--- FAIL: TestDet (0.00s)
    matrix3x3_test.go:355: [ 1.7400 2.7183 42.0000 ] [ 3.1416 1.4142 2.3026 ] [ 3.0000 1.2720 9.8976 ].det() = -56.83852522412309, want -56.838525224123096
--- FAIL: TestPointMeasuresPointArea (0.00s)
    point_measures_test.go:65: PointArea((0.257000000000000006217249, -0.572300000000000030908609, 0.112000000000000002331468), (0.257000000000000006217249, -0.572300000000000030908609, 0.112000000000000002331468), (0.257000000000000006217249, -0.572300000000000030908609, 0.112000000000000002331468)), got 1.9819499942010192e-34 want 0
--- FAIL: TestPredicatesRobustSignEqualities (0.00s)
    predicates_test.go:130: Testing equality for RobustSign. (0.577350269189625731058868, 0.577350269189625731058868, 0.577350269189625731058868) = (0.577350269189625842081171, 0.577350269189625842081171, 0.577350269189625842081171), got false want true
FAIL
FAIL	github.com/blevesearch/geo/s2	3.985s
```

It'd be good if you can address those alongside the github workflow you're about to add here. We can get that merged in first and rebase your sequence of commits over it - so commit validation is run over every pull request.

--- a/geojson/geojson_s2_util.go
+++ b/geojson/geojson_s2_util.go
@@ -25,12 +25,13 @@ import (
 
 // ------------------------------------------------------------------------
 
+// project the point to all of the linestrings and check if
+// any of the projections are equal to the point.
 func polylineIntersectsPoint(pls []*s2.Polyline,
 	point *s2.Point) bool {
-	s2cell := s2.CellFromPoint(*point)
-
 	for _, pl := range pls {
-		if pl.IntersectsCell(s2cell) {
+		closest, _ := pl.Project(*point)
+		if closest.ApproxEqual(*point) {
 			return true
 		}
 	}
@@ -243,20 +244,6 @@ func s2Cap(vertices []float64, radiusInMeter float64) *s2.Cap {
 	angle := radiusInMetersToS1Angle(float64(radiusInMeter))
 	cap := s2.CapFromCenterAngle(cp, angle)
 	return &cap
-}
-
-func max(a, b float64) float64 {
-	if a >= b {
-		return a
-	}
-	return b
-}
-
-func min(a, b float64) float64 {
-	if a >= b {
-		return b
-	}
-	return a
 }
 
 func StripCoveringTerms(terms []string) []string {

--- a/geojson/geojson_shapes_impl.go
+++ b/geojson/geojson_shapes_impl.go
@@ -125,16 +125,14 @@ func (p *Point) Marshal() ([]byte, error) {
 
 func (p *Point) Intersects(other index.GeoJSON) (bool, error) {
 	p.init()
-	s2cell := s2.CellFromPoint(*p.s2point)
 
-	return checkCellIntersectsShape(&s2cell, p, other)
+	return checkPointIntersectsShape(p.s2point, p, other)
 }
 
 func (p *Point) Contains(other index.GeoJSON) (bool, error) {
 	p.init()
-	s2cell := s2.CellFromPoint(*p.s2point)
 
-	return checkCellContainsShape([]*s2.Cell{&s2cell}, other)
+	return checkPointContainsShape([]*s2.Point{p.s2point}, other)
 }
 
 func (p *Point) Coordinates() []float64 {
@@ -205,8 +203,7 @@ func (p *MultiPoint) Intersects(other index.GeoJSON) (bool, error) {
 	p.init()
 
 	for _, s2point := range p.s2points {
-		cell := s2.CellFromPoint(*s2point)
-		rv, err := checkCellIntersectsShape(&cell, p, other)
+		rv, err := checkPointIntersectsShape(s2point, p, other)
 		if rv && err == nil {
 			return rv, nil
 		}
@@ -217,14 +214,13 @@ func (p *MultiPoint) Intersects(other index.GeoJSON) (bool, error) {
 
 func (p *MultiPoint) Contains(other index.GeoJSON) (bool, error) {
 	p.init()
-	s2cells := make([]*s2.Cell, 0, len(p.s2points))
 
-	for _, s2point := range p.s2points {
-		cell := s2.CellFromPoint(*s2point)
-		s2cells = append(s2cells, &cell)
+	rv, err := checkPointContainsShape(p.s2points, other)
+	if rv && err == nil {
+		return rv, nil
 	}
 
-	return checkCellContainsShape(s2cells, other)
+	return false, nil
 }
 
 func (p *MultiPoint) Coordinates() [][]float64 {
@@ -760,11 +756,15 @@ func NewGeoCircle(points []float64,
 	if err != nil {
 		return nil
 	}
-
-	return &Circle{Typ: CircleType,
+	rv := &Circle{
+		Typ:            CircleType,
 		Vertices:       points,
 		Radius:         radius,
-		radiusInMeters: r}
+		radiusInMeters: r,
+	}
+	rv.init()
+
+	return rv
 }
 
 func (c *Circle) Type() string {
@@ -838,7 +838,10 @@ type Envelope struct {
 }
 
 func NewGeoEnvelope(points [][]float64) index.GeoJSON {
-	return &Envelope{Vertices: points, Typ: EnvelopeType}
+	rv := &Envelope{Vertices: points, Typ: EnvelopeType}
+	rv.init()
+
+	return rv
 }
 
 func (e *Envelope) Type() string {
@@ -884,15 +887,13 @@ func (e *Envelope) Contains(other index.GeoJSON) (bool, error) {
 
 //--------------------------------------------------------
 
-// checkCellIntersectsShape checks for intersection between
-// the s2cell and the shape in the document.
-func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
-	other index.GeoJSON) (bool, error) {
+// checkPointIntersectsShape checks for intersection between
+// the point and the shape in the document.
+func checkPointIntersectsShape(point *s2.Point, shapeIn, other index.GeoJSON) (bool, error) {
 	// check if the other shape is a point.
 	if p2, ok := other.(*Point); ok {
-		s2cell := s2.CellFromPoint(*p2.s2point)
-
-		if cell.IntersectsCell(s2cell) {
+		// Check if the points are equal
+		if point.ApproxEqual(*p2.s2point) {
 			return true, nil
 		}
 
@@ -901,11 +902,9 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 
 	// check if the other shape is a multipoint.
 	if p2, ok := other.(*MultiPoint); ok {
-		// check the intersection for any point in the array.
-		for _, point := range p2.s2points {
-			s2cell := s2.CellFromPoint(*point)
-
-			if cell.IntersectsCell(s2cell) {
+		// check if any of the points are equal
+		for _, p := range p2.s2points {
+			if point.ApproxEqual(*p) {
 				return true, nil
 			}
 		}
@@ -915,8 +914,12 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 
 	// check if the other shape is a polygon.
 	if p2, ok := other.(*Polygon); ok {
-
-		if p2.s2pgn.IntersectsCell(*cell) {
+		// check if the point is contained within the polygon.
+		// polygon contains point will consider vertices to be outside
+		// so we create a shape index and query it instead
+		idx := s2.NewShapeIndex()
+		idx.Add(p2.s2pgn)
+		if s2.NewContainsPointQuery(idx, s2.VertexModelClosed).Contains(*point) {
 			return true, nil
 		}
 
@@ -925,12 +928,15 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 
 	// check if the other shape is a multipolygon.
 	if p2, ok := other.(*MultiPolygon); ok {
-		// check the intersection for any polygon in the collection.
+		// check if the point is contained within any of the polygons
+		// polygon contains point will consider vertices to be outside
+		// so we create a shape index and query it instead
+		idx := s2.NewShapeIndex()
 		for _, s2pgn := range p2.s2pgns {
-
-			if s2pgn.IntersectsCell(*cell) {
-				return true, nil
-			}
+			idx.Add(s2pgn)
+		}
+		if s2.NewContainsPointQuery(idx, s2.VertexModelClosed).Contains(*point) {
+			return true, nil
 		}
 
 		return false, nil
@@ -938,13 +944,11 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 
 	// check if the other shape is a linestring.
 	if p2, ok := other.(*LineString); ok {
-		for i := 0; i < p2.pl.NumEdges(); i++ {
-			edge := p2.pl.Edge(i)
-			start := s2.CellFromPoint(edge.V0)
-			end := s2.CellFromPoint(edge.V1)
-			if cell.IntersectsCell(start) || cell.IntersectsCell(end) {
-				return true, nil
-			}
+		// project the point to the linestring and check if
+		// the projection is equal to the point.
+		closest, _ := p2.pl.Project(*point)
+		if closest.ApproxEqual(*point) {
+			return true, nil
 		}
 
 		return false, nil
@@ -954,13 +958,9 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 	if p2, ok := other.(*MultiLineString); ok {
 		// check the intersection for any linestring in the array.
 		for _, pl := range p2.pls {
-			for i := 0; i < pl.NumEdges(); i++ {
-				edge := pl.Edge(i)
-				start := s2.CellFromPoint(edge.V0)
-				end := s2.CellFromPoint(edge.V1)
-				if cell.IntersectsCell(start) || cell.IntersectsCell(end) {
-					return true, nil
-				}
+			closest, _ := pl.Project(*point)
+			if closest.ApproxEqual(*point) {
+				return true, nil
 			}
 		}
 
@@ -979,8 +979,10 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 
 	// check if the other shape is a circle.
 	if c, ok := other.(*Circle); ok {
-
-		if c.s2cap.IntersectsCell(*cell) {
+		// check if the point is contained within the circle
+		// by calculating the distance between the point and the
+		// center of the circle.
+		if c.s2cap.ContainsPoint(*point) {
 			return true, nil
 		}
 
@@ -989,8 +991,9 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 
 	// check if the other shape is an envelope.
 	if e, ok := other.(*Envelope); ok {
-
-		if e.r.IntersectsCell(*cell) {
+		// check if the point is contained by the envelope
+		// by checking if the point is within its bounds
+		if e.r.ContainsPoint(*point) {
 			return true, nil
 		}
 
@@ -1001,15 +1004,14 @@ func checkCellIntersectsShape(cell *s2.Cell, shapeIn,
 		" found in document", other.Type())
 }
 
-// checkCellContainsShape checks whether the given shape in
-// in the document is contained with the s2cell.
-func checkCellContainsShape(cells []*s2.Cell,
+// checkPointContainsShape checks whether the given shape in
+// in the document is approximately contained with the point.
+func checkPointContainsShape(points []*s2.Point,
 	other index.GeoJSON) (bool, error) {
 	// check if the other shape is a point.
 	if p2, ok := other.(*Point); ok {
-		for _, cell := range cells {
-
-			if cell.ContainsPoint(*p2.s2point) {
+		for _, point := range points {
+			if point.ApproxEqual(*p2.s2point) {
 				return true, nil
 			}
 		}
@@ -1018,12 +1020,12 @@ func checkCellContainsShape(cells []*s2.Cell,
 	}
 
 	// check if the other shape is a multipoint, if so containment is
-	// checked for every point in the multipoint with every given cells.
+	// checked for every point in the multipoint with every given point.
 	if p2, ok := other.(*MultiPoint); ok {
 		// check the containment for every point in the collection.
 		lookup := make(map[int]struct{})
-		for _, cell := range cells {
-			for pos, point := range p2.s2points {
+		for _, qpoint := range points {
+			for pos, dpoint := range p2.s2points {
 				if _, done := lookup[pos]; done {
 					continue
 				}
@@ -1032,7 +1034,7 @@ func checkCellContainsShape(cells []*s2.Cell,
 					return true, nil
 				}
 
-				if cell.ContainsPoint(*point) {
+				if qpoint.ApproxEqual(*dpoint) {
 					lookup[pos] = struct{}{}
 				}
 			}

--- a/geojson/geojson_shapes_impl.go
+++ b/geojson/geojson_shapes_impl.go
@@ -946,8 +946,7 @@ func checkPointIntersectsShape(point *s2.Point, shapeIn, other index.GeoJSON) (b
 	if p2, ok := other.(*LineString); ok {
 		// project the point to the linestring and check if
 		// the projection is equal to the point.
-		closest, _ := p2.pl.Project(*point)
-		if closest.ApproxEqual(*point) {
+		if polylineIntersectsPoint([]*s2.Polyline{p2.pl}, point) {
 			return true, nil
 		}
 
@@ -957,11 +956,8 @@ func checkPointIntersectsShape(point *s2.Point, shapeIn, other index.GeoJSON) (b
 	// check if the other shape is a multilinestring.
 	if p2, ok := other.(*MultiLineString); ok {
 		// check the intersection for any linestring in the array.
-		for _, pl := range p2.pls {
-			closest, _ := pl.Project(*point)
-			if closest.ApproxEqual(*point) {
-				return true, nil
-			}
+		if polylineIntersectsPoint(p2.pls, point) {
+			return true, nil
 		}
 
 		return false, nil

--- a/geojson/geojson_shapes_test.go
+++ b/geojson/geojson_shapes_test.go
@@ -1,0 +1,381 @@
+//  Copyright (c) 2025 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package geojson
+
+import (
+	"testing"
+
+	index "github.com/blevesearch/bleve_index_api"
+)
+
+func TestPointIntersects(t *testing.T) {
+
+	tests := []struct {
+		queryPoint *Point
+		other      index.GeoJSON
+		output     bool
+	}{
+		{ // 0 - Same point with 15 decimal places
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234567, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 1 - Point with 15th decimal place differing
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234568, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 2 - Point with 13th decimal place differing
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234667, 1.234567891234567}),
+			output:     false,
+		},
+		{ // 3 - MultiPoint with a match
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonMultiPoint([][]float64{{1.134567891234567, 1.234567891234567}, {1.234567891234567, 1.234567891234567}}),
+			output:     true,
+		},
+		{ // 4 - MultiPoint with no match
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonMultiPoint([][]float64{{1.234567891234567, 1.134567891234567}, {1.134567891234567, 1.234567891234567}}),
+			output:     false,
+		},
+		{ // 5 - Polygon with point on the inside
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0, 0}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}),
+			output:     true,
+		},
+		{ // 6 - Clockwise polygon with point on the outside
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0, 0}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {-1, 1}, {1, 1}, {1, -1}, {-1, -1}}}),
+			output:     false,
+		},
+		{ // 7 - Polygon with point on the vertex
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{-1, -1}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}),
+			output:     true,
+		},
+		{ // 8 - Polygon with point on the edge
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0.5, 1}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}),
+			output:     true,
+		},
+		{ // 9 - Polygon with point in the hole
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0, 0}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}, {{-0.5, -0.5}, {-0.5, 0.5}, {0.5, 0.5}, {0.5, -0.5}, {-0.5, -0.5}}}),
+			output:     false,
+		},
+		{ // 10 - MulitiPolygon with point
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{2.5, 2.5}},
+			other:      NewGeoJsonMultiPolygon([][][][]float64{{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}, {{{2, 2}, {3, 2}, {3, 3}, {2, 3}, {2, 2}}}}),
+			output:     true,
+		},
+		{ // 11 - MultiPolygon without point
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{2.5, 2.5}},
+			other:      NewGeoJsonMultiPolygon([][][][]float64{{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}, {{{-2, -2}, {-3, -2}, {-3, -3}, {-2, -3}, {-2, -2}}}}),
+			output:     false,
+		},
+		{ // 12 - LineString with point on the line
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0, 0}},
+			other:      NewGeoJsonLinestring([][]float64{{-1, 0}, {1, 0}}),
+			output:     true,
+		},
+		{ // 13 - LineString with point on the vertex
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{-1, 0}},
+			other:      NewGeoJsonLinestring([][]float64{{-1, 0}, {1, 0}}),
+			output:     true,
+		},
+		{ // 14 - LineString with point not on line
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{-2, 0}},
+			other:      NewGeoJsonLinestring([][]float64{{-1, 0}, {1, 0}}),
+			output:     false,
+		},
+		{ // 15 - MultiLineString with point on the line
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1, 0}},
+			other:      NewGeoJsonMultilinestring([][][]float64{{{-5, 0}, {-3, 0}}, {{-2, 0}, {2, 0}}}),
+			output:     true,
+		},
+		{ // 16 - MultiLineString with point on the vertex
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{2, 1}},
+			other:      NewGeoJsonMultilinestring([][][]float64{{{-1, 0}, {1, 0}}, {{-2, 1}, {2, 1}}}),
+			output:     true,
+		},
+		{ // 17 - MultiLineString with point not on line
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{-3, 1}},
+			other:      NewGeoJsonMultilinestring([][][]float64{{{-1, 0}, {1, 0}}, {{-2, 1}, {2, 1}}}),
+			output:     false,
+		},
+		{ // 18 - Circle with point not on the inside
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0, 2}},
+			other:      NewGeoCircle([]float64{0, 0}, "1km"),
+			output:     false,
+		},
+		{ // 19 - Circle with point on the inside
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0, 0.03}},
+			other:      NewGeoCircle([]float64{0, 0}, "10km"),
+			output:     true,
+		},
+		{ // 20 - Envelope with point on the inside
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{0, 0}},
+			other:      NewGeoEnvelope([][]float64{{-2, 2}, {2, -2}}),
+			output:     true,
+		},
+		{ // 21 - Envelope with point on the outside
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{3, 2}},
+			other:      NewGeoEnvelope([][]float64{{-2, 2}, {2, -2}}),
+			output:     false,
+		},
+		{ // 22 - Envelope with point on the edge
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1, 2}},
+			other:      NewGeoEnvelope([][]float64{{-2, 2}, {2, -2}}),
+			output:     true,
+		},
+	}
+
+	for i, test := range tests {
+		result, err := test.queryPoint.Intersects(test.other)
+		if err != nil {
+			t.Errorf("Error: %v", err)
+		}
+
+		if result != test.output {
+			t.Errorf("Test - %d, expected %v, got %v", i, test.output, result)
+		}
+	}
+}
+
+func TestMultiPointIntersects(t *testing.T) {
+	tests := []struct {
+		queryPoint *MultiPoint
+		other      index.GeoJSON
+		output     bool
+	}{
+		{ // 0 - Same point with 15 decimal places
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234567, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 1 - Point with 15th decimal place differing
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234568, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 2 - Point with 13th decimal place differing
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234667, 1.234567891234567}),
+			output:     false,
+		},
+		{ // 3 - MultiPoint with a match
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonMultiPoint([][]float64{{1.134567891234567, 1.234567891234567}, {1.234567891234567, 1.234567891234567}}),
+			output:     true,
+		},
+		{ // 4 - MultiPoint with no match
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonMultiPoint([][]float64{{1.234567891234567, 1.134567891234567}, {1.134567891234567, 1.234567891234567}}),
+			output:     false,
+		},
+		{ // 5 - Polygon with point on the inside
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{0, 0}, {4, 4}}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}),
+			output:     true,
+		},
+		{ // 6 - Clockwise polygon with point on the outside
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{0.5, 0.5}, {0, 0}}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {-1, 1}, {1, 1}, {1, -1}, {-1, -1}}}),
+			output:     false,
+		},
+		{ // 7 - Polygon with point on the vertex
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {-1, -1}}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}),
+			output:     true,
+		},
+		{ // 8 - Polygon with point on the vertex
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{-0.5, -1}, {4, 4}}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}),
+			output:     true,
+		},
+		{ // 9 - Polygon with point in the hole
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{0, 0}, {4, 4}}},
+			other:      NewGeoJsonPolygon([][][]float64{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}, {{-0.5, -0.5}, {-0.5, 0.5}, {0.5, 0.5}, {0.5, -0.5}, {-0.5, -0.5}}}),
+			output:     false,
+		},
+		{ // 10 - MulitiPolygon with point
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {0, 0}}},
+			other:      NewGeoJsonMultiPolygon([][][][]float64{{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}, {{{2, 2}, {3, 2}, {3, 3}, {2, 3}, {2, 2}}}}),
+			output:     true,
+		},
+		{ // 11 - MultiPolygon without point
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {-4, -4}}},
+			other:      NewGeoJsonMultiPolygon([][][][]float64{{{{-1, -1}, {1, -1}, {1, 1}, {-1, 1}, {-1, -1}}}, {{{-2, -2}, {-3, -2}, {-3, -3}, {-2, -3}, {-2, -2}}}}),
+			output:     false,
+		},
+		{ // 12 - LineString with point on the line
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{0, 0}, {-1, -1}}},
+			other:      NewGeoJsonLinestring([][]float64{{-1, 0}, {1, 0}}),
+			output:     true,
+		},
+		{ // 13 - LineString with point on the vertex
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1, 0}, {4, 4}}},
+			other:      NewGeoJsonLinestring([][]float64{{-1, 0}, {1, 0}}),
+			output:     true,
+		},
+		{ // 14 - LineString with point not on line
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {2, 3}}},
+			other:      NewGeoJsonLinestring([][]float64{{-1, 0}, {1, 0}}),
+			output:     false,
+		},
+		{ // 15 - MultiLineString with point on the line
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{-2, 0}, {4, 4}}},
+			other:      NewGeoJsonMultilinestring([][][]float64{{{-5, 0}, {-3, 0}}, {{-2, 0}, {2, 0}}}),
+			output:     true,
+		},
+		{ // 16 - MultiLineString with point on the vertex
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {-2, 1}}},
+			other:      NewGeoJsonMultilinestring([][][]float64{{{-1, 0}, {1, 0}}, {{-2, 1}, {2, 1}}}),
+			output:     true,
+		},
+		{ // 17 - MultiLineString with point not on line
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1, -1}, {4, 4}}},
+			other:      NewGeoJsonMultilinestring([][][]float64{{{-1, 0}, {1, 0}}, {{-2, 1}, {2, 1}}}),
+			output:     false,
+		},
+		{ // 18 - Circle with point not on the inside
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {-1, -3}}},
+			other:      NewGeoCircle([]float64{0, 0}, "1km"),
+			output:     false,
+		},
+		{ // 19 - Circle with point on the inside
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{0.024, -0.037}, {4, 4}}},
+			other:      NewGeoCircle([]float64{0, 0}, "10km"),
+			output:     true,
+		},
+		{ // 20 - Envelope with point on the inside
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {0, 0}}},
+			other:      NewGeoEnvelope([][]float64{{-2, 2}, {2, -2}}),
+			output:     true,
+		},
+		{ // 21 - Envelope with point on the outside
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{-2, -3}, {4, 4}}},
+			other:      NewGeoEnvelope([][]float64{{-2, 2}, {2, -2}}),
+			output:     false,
+		},
+		{ // 22 - Envelope with point on the edge
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{4, 4}, {-1, -2}}},
+			other:      NewGeoEnvelope([][]float64{{-2, 2}, {2, -2}}),
+			output:     true,
+		},
+	}
+
+	for i, test := range tests {
+		result, err := test.queryPoint.Intersects(test.other)
+		if err != nil {
+			t.Errorf("Error: %v", err)
+		}
+
+		if result != test.output {
+			t.Errorf("Test - %d, expected %v, got %v", i, test.output, result)
+		}
+	}
+}
+
+func TestPointContains(t *testing.T) {
+	tests := []struct {
+		queryPoint *Point
+		other      index.GeoJSON
+		output     bool
+	}{
+		{ // 0 - Same point with 15 decimal places
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234567, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 1 - Point with 15th decimal place differing
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234568, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 2 - Point with 13th decimal place differing
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234667, 1.234567891234567}),
+			output:     false,
+		},
+		{ // 3 - MultiPoint with a match
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonMultiPoint([][]float64{{1.234567891234567, 1.234567891234567}}),
+			output:     true,
+		},
+		{ // 4 - MultiPoint with no match
+			queryPoint: &Point{Typ: PointType, Vertices: []float64{1.234567891234567, 1.234567891234567}},
+			other:      NewGeoJsonMultiPoint([][]float64{{1.234567891234567, 1.134567891234567}, {1.134567891234567, 1.234567891234567}}),
+			output:     false,
+		},
+	}
+
+	for i, test := range tests {
+		result, err := test.queryPoint.Contains(test.other)
+		if err != nil {
+			t.Errorf("Error: %v", err)
+		}
+
+		if result != test.output {
+			t.Errorf("Test - %d, expected %v, got %v", i, test.output, result)
+		}
+	}
+}
+func TestMultiPointContains(t *testing.T) {
+	tests := []struct {
+		queryPoint *MultiPoint
+		other      index.GeoJSON
+		output     bool
+	}{
+		{ // 0 - Same point with 15 decimal places
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234567, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 1 - Point with 15th decimal place differing
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234568, 1.234567891234567}),
+			output:     true,
+		},
+		{ // 2 - Point with 13th decimal place differing
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonPoint([]float64{1.234567891234667, 1.234567891234567}),
+			output:     false,
+		},
+		{ // 3 - MultiPoint with a match
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonMultiPoint([][]float64{{2.234567891234567, 2.234567891234567}, {1.234567891234567, 1.234567891234567}}),
+			output:     true,
+		},
+		{ // 4 - MultiPoint with no match
+			queryPoint: &MultiPoint{Typ: PointType, Vertices: [][]float64{{1.234567891234567, 1.234567891234567}, {2.234567891234567, 2.234567891234567}}},
+			other:      NewGeoJsonMultiPoint([][]float64{{1.234567891234567, 1.134567891234567}, {1.134567891234567, 1.234567891234567}}),
+			output:     false,
+		},
+	}
+
+	for i, test := range tests {
+		result, err := test.queryPoint.Contains(test.other)
+		if err != nil {
+			t.Errorf("Error: %v", err)
+		}
+
+		if result != test.output {
+			t.Errorf("Test - %d, expected %v, got %v", i, test.output, result)
+		}
+	}
+}

